File: src/main/java/fr/spoutnik87/musicbot_rest/model/Media.java
package fr.spoutnik87.musicbot_rest.model;

import com.fasterxml.jackson.annotation.JsonManagedReference;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonView;
import lombok.Data;
import lombok.NoArgsConstructor;
import lombok.NonNull;
import lombok.RequiredArgsConstructor;

import javax.persistence.*;
import java.io.Serializable;

@NoArgsConstructor
@RequiredArgsConstructor
@Data
@Entity
@Table(name = "Media")
public class Media extends AuditModel implements Serializable {

  @JsonView(Views.Public.class)
  @NonNull
  @JsonProperty("id")
  @Column(nullable = false, unique = true)
  private String uuid;

  @JsonView(Views.Public.class)
  @NonNull
  @Column(nullable = false)
  private String name;

  @JsonView(Views.Public.class)
  @NonNull
  @ManyToOne
  @JoinColumn(name = "media_type_id")
  @JsonManagedReference
  private MediaType mediaType;

  @JsonView(Views.Public.class)
  @NonNull
  @ManyToOne
  @JoinColumn(name = "category_id")
  @JsonManagedReference
  private Category category;
}

File: src/main/kotlin/fr/spoutnik87/musicbot_rest/model/Media.kt
package fr.spoutnik87.musicbot_rest.model

import java.io.Serializable
import javax.persistence.*

@Entity
@Table(name = "Media")
data class Media(
        @Column(nullable = false, unique = true)
        var uuid: String,
        @Column(nullable = false)
        var name: String
) : AuditModel(), Serializable {

    var extension: String? = null

    var size: Long? = null

    @Column(nullable = false)
    var content: Boolean = false

    @Column(nullable = false)
    var thumbnail: Boolean = false

    @ManyToOne
    @JoinColumn(name = "media_type_id")
    lateinit var mediaType: MediaType

    @ManyToOne
    @JoinColumn(name = "category_id")
    lateinit var category: Category

    @OneToMany(mappedBy = "media", cascade = [CascadeType.ALL])
    val mediaGroupSet: MutableSet<MediaGroup> = HashSet()

    val groupList
        get() = mediaGroupSet.map { it.group }

    constructor(uuid: String, name: String, mediaType: MediaType, category: Category) : this(uuid, name) {
        this.mediaType = mediaType
        this.category = category
    }

    fun hasContent() = content

    fun hasThumbnail() = thumbnail
}